proposals/NNNN-add-result.md

```diff
@@ -0,0 +1,242 @@
+
+## Motivation
+
+Swift's [Error Handling Rationale and Proposal](https://github.com/apple/swift/blob/master/docs/ErrorHandlingRationale.rst) document lays out the reasoning behind and high level details of the current Swift error handling story. Types conforming to `Error` can be propagated and handled using the `do` `try` `catch` `throw` syntax. The rationale document refers to this model as a typed and automatically propagating error system. However, this system has several drawbacks, some of which are mentioned in the rationale document. Namely, it cannot compose with asynchronous work, more complex error handling, or with failure values which don't conform to `Error`. The added flexibility of typed, marked, but manually propagating error type can address these shortcomings. Namely `Result<Value, Error>`.
+
+## Proposed solution
+
+```swift
+public enum Result<Value, Error> {
+    case success(Value), failure(Error)
+}
+```
+
+`Result<Value, Error>` is a pragmatic compromise between competing error handling concerns both present and future. It is unconstrained for the `Error` type captured in the `.failure` case to both allow for a possible type-`throws` future and allow failures to return values that don't conform to `Swift.Error`.
+
+### Usage
+
+#### Asynchronous APIs
+
+Most commonly, and seen in abundance when using Apple or Foundation APIs, `Result<Value, Error>` can serve to unify the awkwardly disparate parameters seen in asynchronous completion handlers. For instance, `URLSession`'s completion handlers take three optional parameters:
+
+```swift
+func dataTask(with url: URL, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
+```
+
+This can make it quite difficult to elegantly consume the results of these APIs:
+
+```swift
+URLSession.shared.dataTask(with: url) { (data, response, error) in
+    guard error != nil else { self.handleError(error!) }
+    
+    guard let data = data, let response = response else { return // Impossible? }
+    
+    handleResponse(response, data: data)
+}
+```
+
+While this code is only a few lines long, it exposes Swift's complete lack of automatic error handling for asynchronous APIs. Not only was the `error` forcibly unwrapped (or perhaps handled using a slightly less elegant `if` statement), but a possibly impossible scenario was created. What happens if `response` or `data` are `nil`? Is it even possible? It shouldn't be, but Swift currently lacks the ability to express this impossibility. Using `Result<Value, Error>` for the same scenario allows for much more elegant code:
+
+```swift
+URLSession.shared.dataTask(with: url) { (result: Result<(Data, URLResponse), (Error, URLResponse?)>) in // Type added for illustration purposes.
```

Comment:
Unfortunately, this contradicts the implementation of `init(_ throwing: () throws -> Value)` below because you cannot throw `(Error, URLResponse?)`.

Reply:
I've updated the implementation in the proposal to match the current state of the PR, which places the throwing initializer in a constrained extension.